### ISSUE TITLE
Constraint changes causes replacements resulting in units being destroyed

### ISSUE THREAD

**Issue by hloeung** [juju/terraform-provider-juju#632]
### Description

Constraint changes or differences (see https://github.com/juju/terraform-provider-juju/issues/344) causes units to be replaced and destroyed. This is scary for an environment hosting databases.

I can see use cases where we want to increase or scale up DB units. We do so by changing the constraints, adding new units, then cycling out the old.

Sadly, the current Juju Terraform provider sees constraint changes and causes replacements which means destroying units.

### Urgency

Casually reporting

### Terraform Juju Provider version

0.15.0

### Terraform version

v1.7.2-dev

### Juju version

3.5.3

### Terraform Configuration(s)

```terraform
locals {
  # Work around Juju `unexpected new value: .constraints: was cty.StringVal("... root-disk=20G"), but now cty.StringVal(" ... root-disk=20480M").`
  # https://github.com/juju/terraform-provider-juju/issues/344
  postgresql_disk_gigabytes       = var.disk_gigabytes * 1024
  postgresql_charm_constraints    = "arch=amd64 cores=${var.cores} mem=${var.mem_megabytes}M root-disk=${local.postgresql_disk_gigabytes}M"
  postgresql_root_disk_constraint = var.postgresql_use_local_storage ? " root-disk-source=local" : ""
}

resource "juju_application" "postgresql" {
  name  = "postgresql"
  model = var.juju_model_name
  trust = true

  charm {
    name     = "postgresql"
    channel  = var.postgresql_charm_channel
    revision = var.postgresql_charm_revision
    base     = var.postgresql_charm_base
  }

  expose {}

  units       = var.postgresql_charm_units
  constraints = "${local.postgresql_charm_constraints}${local.postgresql_root_disk_constraint}"
}
```

### Reproduce / Test

```shell
terraform init & terraform apply

Update `postgresql_disk_gigabytes` to some other value

terraform apply
```

### Debug/Panic Output

```shell
# module.postgresql-database.juju_application.postgresql must be replaced
-/+ resource "juju_application" "postgresql" {
      ~ constraints = "arch=amd64 cores=2 mem=4096M root-disk=20480M" -> "arch=amd64 cores=2 mem=4096M root-disk=30720M" # forces replacement
      ~ id          = "stg-is-dbaas-testing:postgresql" -> (known after apply)
        name        = "postgresql"
      ~ placement   = "5,6,7" -> (known after apply)
      + principal   = (known after apply)
      ~ storage     = [
          - {
              - count = 1 -> null
              - label = "pgdata-3" -> null
              - pool  = "rootfs" -> null
              - size  = "20G" -> null
            },
          - {
              - count = 1 -> null
              - label = "pgdata-4" -> null
              - pool  = "rootfs" -> null
              - size  = "20G" -> null
            },
          - {
              - count = 1 -> null
              - label = "pgdata-5" -> null
              - pool  = "rootfs" -> null
              - size  = "20G" -> null
            },
        ] -> (known after apply)
        # (3 unchanged attributes hidden)

      ~ charm {
            name     = "postgresql"
          ~ series   = "jammy" -> (known after apply)
            # (3 unchanged attributes hidden)
        }

        # (1 unchanged block hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```
```


### Notes & References

_No response_

**Comment by hmlanigan** [comment]
The juju cli does not allow for constraints of existing units to be updated either. If you change an application's constraints, they apply to the next machines for the units deployed only, see [juju set-constraints](https://juju.is/docs/juju/juju-set-constraints).

Juju lacks the ability to resize existing machines disk today.

As terraform is a declarative language so the expectation would be that every unit has machines with the same constraints. It's not as flexible as juju. 

Your next problem would be that the provider will not let you choose which units are to be deleted. The oldest units are removed first.

I'd suggest that putting the DB on the root disk is a bad idea anyways. You'd want it on a volume which can be detached and reattached to specific units (`juju attach-storage`). This is not currently supported by the provider.

If this is a real world scenario, a design of how to accomplish it should be made and agreed to.

**Comment by hloeung** [comment]
Yes this is a real world scenario - databases grow and eventually needs to be resized increasing storage. Even with Juju managed storage (`juju attach-storage`), there isn't a way currently to grow the existing storage.

There's also the vCPU and RAM - as applications become used more, I can see a scenario where we need to scale up the no. of vCPUs and RAM.

We're fine with the Terraform provider not manage which units are deleted and which to cycle out. We just want a way to be able to update the constraint in the Terraform plan, then manually cycle out units and have Terraform provision new units in it's place.

**Comment by hloeung** [comment]
Another use case is that constraints can be instance types, e.g. for Azure, `constraints: instance-type=Standard_D3_v2`. Instance types can be retired, as recently Azure announced the retirement of the Dv2 series[1].

For these workloads, we want to be able to update the Terraform plan to provision new units using the new instance type chosen and slowly age out or retire the existing. `instance-type=Standard_D3_v4`

[1]: https://learn.microsoft.com/en-us/azure/virtual-machines/migration/sizes/d-ds-dv2-dsv2-ls-series-migration-guide

**Comment by kian99** [comment]
@hloeung After some discussion these are the main ideas,
- Because Terraform is declarative, it is intentional that when you look at a plan, your infra is a copy of that plan. The `alwaysMatchConstraints` opens a can of worms where we violate that invariant.
- If there is a scenario where someone wants to update constraints without triggering a replacement, it's a bit outside the scope of Terraform to provide for this.

For your scenario to "slowly age out or retire the existing" units, it's a bit hard if you do this over a long period of time. Ideally if this were to be done in, let's say a day, you could:
- Use Juju CLI, update application constraints
- Bring in new units with the new constraints
- Destroy old units with old constraints
- Import application into TF state / Manually update Terraform's state

If you pair this approach with Terraform's [ignore_changes](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#ignore_changes) directive, you could at least continue to run Terraform operations while the constraint doesn't match.

Let me know your thoughts on this.

**Comment by kian99** [comment]
Here's an example from the [official docs](https://developer.hashicorp.com/terraform/language/block/resource#ignore-attribute-changes), and a specific example of an application resource,

```terraform
resource "juju_model" "this" {
  name = "test-model"
}

resource "juju_application" "app" {
  name       = "app"
  model_uuid = juju_model.this.uuid

  charm {
    name     = "apache2"
    channel  = "latest/stable"
    revision = 64
    base     = "ubuntu@22.04"
  }

  constraints = "mem=2G cores=4"

  lifecycle {
    ignore_changes = [constraints]
  }
}
```
With this, Terraform will detect no changes if you change the plan or if you change the application constraints with the Juju CLI.

**Comment by hloeung** [comment]
Thanks - closing this off in favor of the suggested lifecycle ignore constraint changes.
